geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/MinMax.scala

```diff
@@ -23,114 +24,162 @@ import scala.reflect.ClassTag
+    if (isEmpty) {
+      """{ "min": null, "max": null, "cardinality": 0 }"""
+    } else {
+      s"""{ "min": ${jsonStringify(minValue)}, "max": ${jsonStringify(maxValue)}, "cardinality": $cardinality }"""
+    }
   }
 
-  override def isEmpty: Boolean = min == defaults.min
+  override def isEmpty: Boolean = minValue == defaults.max
 
   override def clear(): Unit = {
-    min = defaults.min
-    max = defaults.max
+    minValue = defaults.max
+    maxValue = defaults.min
+    hpp = new HyperLogLog(10)
   }
 
-  override def equals(other: Any): Boolean = other match {
-    case that: MinMax[T] => attribute == that.attribute && min == that.min && max == that.max
+  override def isEquivalent(other: Stat): Boolean = other match {
+    case that: MinMax[T] =>
+      attribute == that.attribute && minValue == that.minValue &&
+          maxValue == that.maxValue && cardinality == that.cardinality
     case _ => false
   }
-
-  override def hashCode(): Int =
-    Seq(attribute, min, max).map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
 }
 
 object MinMax {
 
   trait MinMaxDefaults[T] {
     def min: T
     def max: T
-    def minmax(value: T, min: T, max: T): (T, T)
+    def min(left: T, right: T): T
+    def max(left: T, right: T): T
   }
 
-  abstract class ComparableMinMax[T <: Comparable[T]] extends MinMaxDefaults[T] {
-    override def minmax(value: T, min: T, max: T): (T, T) = {
-      val mn = if (value.compareTo(min) > 0) min else value
-      val mx = if (value.compareTo(max) < 0) max else value
-      (mn, mx)
+  object MinMaxDefaults {
+    def apply[T](binding: Class[_]): MinMaxDefaults[T] = {
+      if (binding == classOf[String]) {
+        MinMaxString.asInstanceOf[MinMaxDefaults[T]]
+      } else if (binding == classOf[Integer]) {
+        MinMaxInt.asInstanceOf[MinMaxDefaults[T]]
+      } else if (binding == classOf[java.lang.Long]) {
+        MinMaxLong.asInstanceOf[MinMaxDefaults[T]]
+      } else if (binding == classOf[java.lang.Float]) {
+        MinMaxFloat.asInstanceOf[MinMaxDefaults[T]]
+      } else if (binding == classOf[java.lang.Double]) {
+        MinMaxDouble.asInstanceOf[MinMaxDefaults[T]]
+      } else if (classOf[Date].isAssignableFrom(binding)) {
+        MinMaxDate.asInstanceOf[MinMaxDefaults[T]]
+      } else if (classOf[Geometry].isAssignableFrom(binding)) {
+        MinMaxGeometry.asInstanceOf[MinMaxDefaults[T]]
+      } else {
+        throw new IllegalArgumentException(s"No implicit default available for type: $binding")
+      }
     }
   }
 
+  abstract class ComparableMinMax[T <: Comparable[T]] extends MinMaxDefaults[T] {
+    override def min(left: T, right: T): T = if (left.compareTo(right) > 0) right else left
+    override def max(left: T, right: T): T = if (left.compareTo(right) < 0) right else left
+  }
+
   implicit object MinMaxString extends ComparableMinMax[String] {
-    override val min: String = "~~~"
-    override val max: String = ""
+    override val min: String = ""
```

Comment:
oh, good point. I thought ~ was pretty much the last char, as we use it in accumulo ranges. I may be incorrect about that...